### DISCUSSION TITLE
I am trying to getfeeformessage in SOLANA

### DISCUSSION THREAD

**Discussion by blukhi99** [trustwallet/wallet-core#2619] [Q&A] [answered]
I am trying to [getfeeformessage](https://docs.solana.com/developing/clients/jsonrpc-api#getfeeformessage) in the Solana chain where I need Base64 of encoding message.

I created code  like below

```ts
val transferMessage = Solana.Transfer.newBuilder().apply {
    recipient = AnyAddress(receiver, CoinType.SOLANA).description()
     value = amount.toBigDecimal().toWEI(decimals).toLong()
}.build()
val signingInput = Solana.SigningInput.newBuilder().apply {
    transferTransaction = transferMessage
    recentBlockhash = blockhash
    privateKey = ByteString.copyFrom(Base58.decodeNoCheck(_privateKey))
}.build()
val output = AnySigner.sign(signingInput, CoinType.SOLANA, Solana.SigningOutput.parser())
```
output.encode what I get but, when Pass Base64 encode or without encode Solana throw error  `**"message": "io error: failed to fill whole buffer"**`

Kindly help to resolve this

**Comment by miloserdow** [comment]
Hi, I think what you need to pass as a parameter is `transferMessage` encoded with Base64, not the signed transaction

**Comment by blukhi99** [comment]
Thanks for answer
 
I also thought but that is also not working

**Comment by Milerius** [comment]
> Thanks for answer
> 
> I also thought but that is also not working

Hello @blukhi99 - can you give more information, what's the bug you are facing now?

**Comment by optout21** [comment]
Some comments:
- `SigningOutput.encoded` is Base58-encoded; `getfeeformessage` expects Base64-encoded
- `SigningOutput.encoded` contains the signatures also, maybe `getfeeformessage` expects only the message data (not clear to me from the description); see `Transaction.messageData()` 
https://github.com/trustwallet/wallet-core/blob/master/src/Solana/Transaction.h#L501

**Comment by blukhi99** [comment]
Hello @Milerius  and @catenocrypt 
Thanks for the replay.

let me explain more.
I want to integrate Solana into My android app and I am using a wallet-core library that has some kinds of method already defined for PrivateKey, and Sign.
in other SDK [Solanaj](https://github.com/skynetcap/solanaj) I can see it is working perfectly and for the get Serialize Transaction message first Sign by the private key and use method of serializeMessage and its working 100%.
but, I want to use the Wallet Core library due to the planning of multi-chain integration.

one, more thing I saw Sign Message of Wallet-core lib and Solanj are different for the same parameters.
The error that I get when passing only Base64 encoded message


`{
  "id":1,
  "jsonrpc":"2.0",
  "method":"getFeeForMessage",
  "params":[
    "CixBTFZoaXpHUXU1THU0eHNpam9DOXBQNEhkYjhkN1ZiS1g4MlRnOFpjeHp3aRCgjQY=",
    {
      "commitment":"processed"
    }
  ]
}`


```
{
    "jsonrpc": "2.0",
    "error": {
        "code": -32602,
        "message": "io error: failed to fill whole buffer"
    },
    "id": 1
}
```

**Comment by optout21** [comment]
Do you have an example of a `getFeeForMessage` call that does work? That would be helpful to understand the difference.

**Reply by blukhi99** [reply]
@catenocrypt 
```
Request ->

{
  "id":1,
  "jsonrpc":"2.0",
  "method":"getFeeForMessage",
  "params":[
    "AQABA0bPP47lGq78NYfRyGgqoO/JQSzc9e+tmI81886gzUYbirer241oHT7trWZexeTiNP0R5E+F7Vmy+WtN1iVJenEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAWMOVRIG/M4hJDWAGuA9tgTG+2ydWsZJGLaH5+QQQ7AQICAAEMAgAAAAEAAAAAAAAA",
    {
      "commitment":"processed"
    }
  ]
}

Response ->

{
    "jsonrpc": "2.0",
    "result": {
        "context": {
            "apiVersion": "1.10.39",
            "slot": 153762377
        },
        "value": 5000
    },
    "id": 1
}
```

**Comment by optout21** [comment]
The data in your example seems incorrect (`CixBTFZoaXpHUXU1THU0eHNpam9DOXBQNEhkYjhkN1ZiS1g4MlRnOFpjeHp3aRCgjQY=`).
It seems it is a Base64 encoding of a Base58 encoded data.
As I said, `SigningOutput.encoded` is Base58-encoded, you need to Base58 decode if you need in other format.
Also, this test data seems a bit too short, it includes only one 32-byte address (`ALVhizGQu5Lu4xsijoC9pP4Hdb8d7VbKX82Tg8Zcxzwi`?).
Something is very mixed up.

**Reply by blukhi99** [reply]
@catenocrypt 

Input Data //  Before Sign

Base64 Of InputData = CiAcCH2tNBu-UEV5Fc5D1iFAE_Nni1FrKV7JHUTkeY2ZZBIsQjVpQ1M4NnBrU0NXV1I5VTZYaUNqZ1ptS2FNTHRNbWdqYTk0d3diQVFrdnMaMgosQUxWaGl6R1F1NUx1NHhzaWpvQzlwUDRIZGI4ZDdWYktYODJUZzhaY3h6d2kQwIQ9   

Error = "message": "InvalidByte(11, 45)"

After Sign

outputEncoded = 6HLjNyD5treGWAreCdQrbLvXBQ5tgKZgwzsKXVdShqHdtPtSE4XepHrzRDiWJd63tTPiWvWVUGxQJy9TvFhcdvxcAQZqSXaf2cLNKX335Ex9DeLPyXUZ9mB3cxUNaVuCW7LLbfNtoeCYtRHP58ER4awmXZbW9WpvDfaMFGSQfHcjZfkmoY9BgTFtKF9a2yAuYNeZCsy5oP1cvFfnkcJdMk9HWmAjFMjM8LJFRu8AjotovtKBrLG6xR8pzvdfctXcJL5rtWATeB47MgqCHGZ6M8GYf6J31et9g4ih5


**Base64Utils.encodeUrlSafeNoPadding(Base58.decodeNoCheck(output.encoded))** =

Aeg-oLi8yFW5ltWhcbigJ3DeDAK-602ZGRHfMOr5vS0kaoRkDH9kt19xc0Z3GLGELsgcED2KGfHw53SAe3-YXQ0BAAEDRs8_juUarvw1h9HIaCqg78lBLNz1762YjzXzzqDNRhuKt6vbjWgdPu2tZl7F5OI0_RHkT4XtWbL5a03WJUl6cQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAlcnGNEya-0oVPVjM4o5-yoAs2TIYZEJBOeE7UzswzygBAgIAAQwCAAAAQEIPAAAAAAA

Error = ""message": "InvalidByte(3, 45)"

**Comment by optout21** [comment, marked as answer]
My random guess is that this method expects only message, without signatures. `SignatureOutput.encoded` from wallet core starts out with the signature.

```
01 // 1 signature follows
e83ea0b8bcc855b996d5a171b8a02770de0c02beeb4d991911df30eaf9bd2d246a84640c7f64b75f7173467718b1842ec81c103d8a19f1f0e774807b7f985d0d
01 // signatures
00 // credit-only signed accounts
01 // credit-only unsigned accounts
03 // accounts
46cf3f8ee51aaefc3587d1c8682aa0efc9412cdcf5efad988f35f3cea0cd461b
8ab7abdb8d681d3eedad665ec5e4e234fd11e44f85ed59b2f96b4dd625497a71
0000000000000000000000000000000000000000000000000000000000000000
95c9c6344c9afb4a153d58cce28e7eca802cd9321864424139e13b533b30cf28 // recent blockhash
01 // instructions
02 02 0001 0c 0200000040420f
```

**Reply by blukhi99** [reply]
https://github.com/trustwallet/wallet-core/discussions/2619#discussioncomment-3933382

**Comment by Otlichnick** [comment]
@blukhi99 did you find the way to receive correct message?

**Reply by blukhi99** [reply]
no not yet

**Reply by Otlichnick** [reply]
We can use this method `https://docs.solana.com/developing/clients/jsonrpc-api#getfeecalculatorforblockhash ` as a temporary solution, but this method is deprecated(

**Reply by blukhi99** [reply]
better is to use the perfect solution I saw this method before but due to deprecated did not try

**Comment by korinsek** [comment]
You need a transaction message without signatures for it. Trustwallet-core lib has this in C but it is not exposed. One way is that they expose it :)

The second way is more workaround that. You can remove signatures out of signed transaction message. This will help you understand the format of the transaction https://docs.solana.com/developing/programming-model/transactions#transaction-format

Here is example how you can do it in kotlin:
```
    /**
     * Creates transaction message, used for estimating fee. Remove signature data from the signed Solana transaction message.
     *
     * @return transaction message in Base64 encoding.
     */
    fun createTransactionMessage(signedTransactionMessage: String): String {
        val messageBytes = Base64.decode(signedTransactionMessage, 64)

        // Decode number of signatures from first 3 bytes
        val signatureCount = messageBytes.sliceArray(IntRange(0, 2)).toUByteArray().let { u16Decoder.decode(it) }

        // Remove signatures from transaction message
        return messageBytes.sliceArray(IntRange((signatureCount + signatureCount * 64u).toInt(), messageBytes.size - 1))
            .let {
                Base64.encodeToString(it, 64)
            }
    }
```
Implementation of u16Decoder can be based (rewritten in kotlin) on this rust implementation https://github.com/solana-labs/solana/blob/b8837c04ec3976c9c16d028fbee86f87823fb97f/sdk/program/src/short_vec.rs#L131

**Reply by optout21** [reply]
Thanks for the analysis.
New issue created for this:  https://github.com/trustwallet/wallet-core/issues/2666

**Reply by blukhi99** [reply]
@korinsek 
when I try to decode it throws an error as does not right the Base64 format

6JZLGesXKa22xv6sLU5fEzcjn1sTeaQJnaDsY5fATak2yMncqzABVpmB1PgQ3uEtfLTHh6DPzmRxFz62m3pEZGCvQGztbwWMqKuH9g9xYg6Ps84uicCdHshw2SvDumtT11KbjXd85fJ4B1SV9vKmsf3aA2wJoiGzaKMK7si9JXSpGQPTrZXzHt5XDpLTyPegWBhbnN6cYY6r7hwszLXebmoLPDFTxfNZmKPCCS6AyYPZxKe245xGSmLSCje6r1eAEW8h5QvJKmGgiFiek8LXNQ1o82kM5TrMibbxb


This is my encoded output

![image](https://user-images.githubusercontent.com/21260525/197208463-1f21b285-921f-4a59-b039-a3209de49a4a.png)

**Reply by korinsek** [reply]
@blukhi99 As @catenocrypt wrote this should be now resolved because it looks like they exposed unsigned transaction. Check this PR https://github.com/trustwallet/wallet-core/pull/2699

**Reply by blukhi99** [reply]
@catenocrypt 

Thanks for helping for adding in Test Case but on the android side, there is no JNI method so, does not able to use that.
can you  add at AnyPlanner or any other method?

**Reply by optout21** [reply]
The extra field is in the .proto object, which is 100% accessible from Android side as well.

**Reply by blukhi99** [reply]
Thanks a Lot @catenocrypt it is working nice
